rust/core-lib/src/movement.rs

```diff
@@ -80,20 +84,49 @@ fn vertical_motion(r: SelRegion, view: &View, text: &Rope, line_delta: isize,
     // This code is quite careful to avoid integer overflow.
     // TODO: write tests to verify
     let line = view.line_of_offset(text, active);
+    // Line Length include the newline, we want the non newline count.
+    let mut line_length = view.offset_of_line(text, line.saturating_add(1)) - view.offset_of_line(text, line) - 1;
     if line_delta < 0 && (-line_delta as usize) > line {
-        return (0, Some(col));
+        if enforce_horiz_pos {
+            return (active, Some(col));
+        } else {
+            return (0, Some(col));
+        }
     }
-    let line = if line_delta < 0 {
+    let mut line = if line_delta < 0 {
         line - (-line_delta as usize)
     } else {
         line.saturating_add(line_delta as usize)
     };
     let n_lines = view.line_of_offset(text, text.len());
-    if line > n_lines {
-        return (text.len(), Some(col));
+
+    if !enforce_horiz_pos {
+        if line > n_lines {
+            return (text.len(), Some(col));
+        }
+
+        return (view.line_col_to_offset(text, line, col), Some(col));
+    }
+
+    // If the active columns is longer than the current line, use the current line length.
+    let col = if line_length < col { line_length } else { col };
```

Comment:
@ryanbloom I did this to ensure the column is never longer then the starting position. I wasn't sure of a better way to get the current cursor's horizontal position. This works though it may not be optimal.